### PR TITLE
[Performance] Don't go into main loop if it is last period for vesting and unlock

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
@sjoshisupra  I think this is trickier than I expected since I need to write separate test cases to cover whether the remaining amount is divisible or not. I have the implementation passed all the test cases now. I will write additional test cases later.

File: aptos-move/framework/aptos-stdlib/sources/fixed_point64.move [hunk 1]
```diff
@@ -56,6 +56,19 @@ module aptos_std::fixed_point64 {
...
 
+    /// Returns x / y. The result cannot be greater than MAX_U128.
+    public fun divide(x: FixedPoint64, y: FixedPoint64): FixedPoint64 {
+        let x_raw = get_raw_value(x);
+        let y_raw = get_raw_value(y);
+        // If it is divisable, return the result. If not, return the result + 1.
+        let result = (x_raw as u256) / (y_raw as u256);
+        if ((x_raw as u256) % (y_raw as u256) != 0) {
+            result = result + 1;
+        };
```
Comment (sjoshisupra) [inline aptos-move/framework/aptos-stdlib/sources/fixed_point64.move]: Why is this logic needed here? When you divide two fixed point values and want to return a fixed point value, no scaling is needed. I don't see any reason why if the remainder is non-zero, we should add `1` to the result.

Comment (sjoshisupra) [reply aptos-move/framework/aptos-stdlib/sources/fixed_point64.move]: Also check that divisor is non zero before performing the division.

File: aptos-move/framework/supra-framework/sources/vesting_without_staking.move [hunk 2]
```diff
@@ -495,25 +495,49 @@ module supra_framework::vesting_without_staking {
...
+            // Determine how many periods is needed based on the left_amount
+            let added_fraction = fixed_point32::multiply_u64_return_fixpoint32(last_completed_period - next_period_to_vest + 1, final_fraction);
+            // If the added_fraction is greater than or equal to the left_amount, then we can vest all the left_amount
+            let periods_need =
+                if (fixed_point32::multiply_u64(vesting_record.init_amount, added_fraction) >= vesting_record.left_amount){
+                     if (vesting_record.left_amount % fixed_point32::multiply_u64(vesting_record.init_amount, final_fraction) == 0) {
+                        vesting_record.left_amount / fixed_point32::multiply_u64(vesting_record.init_amount, final_fraction)
+                    } else {
+                        vesting_record.left_amount / fixed_point32::multiply_u64(vesting_record.init_amount, final_fraction) + 1
+                    }
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/vesting_without_staking.move]:
```suggestion
        if(last_completed_period >= next_period_to_vest && vesting_record.left_amount > 0) {
            let final_fraction = *vector::borrow(schedule, vector::length(schedule) - 1);
            let final_fraction_amount = fixed_point32::multiply_u64(vesting_record.init_amount, final_fraction)
            // Determine how many periods is needed based on the left_amount
            let added_fraction = fixed_point32::multiply_u64_return_fixpoint32(last_completed_period - next_period_to_vest + 1, final_fraction);
            // If the added_fraction is greater than or equal to the left_amount, then we can vest all the left_amount
            let periods_need =
                if (fixed_point32::multiply_u64(vesting_record.init_amount, added_fraction) >= vesting_record.left_amount){
                let result =  vesting_record.left_amount / final_fraction_amount;
                    // check if `left_amount` is perfectly divisible by `final_fraction_amount`
                      if (vesting_record.left_amount == final_fraction_amount*result) {
                       result
                    } else {
                       result + 1
                    }
```

File: aptos-move/framework/aptos-stdlib/sources/fixed_point64.move [hunk 3]
```diff
@@ -56,6 +56,19 @@ module aptos_std::fixed_point64 {
...
+        assert!(result <= MAX_U128, ERATIO_OUT_OF_RANGE);
+        create_from_raw_value((result as u128))
+    }
```
Comment (sjoshisupra) [inline aptos-move/framework/aptos-stdlib/sources/fixed_point64.move]: I don't think this division is correct. If both numbers have 64 bit allotted for fractional value, after division, the fraction would disappear. I believe `x_raw << 64` would be required before performing the division to ensure `result` still has 64 bit fractional bits.

**Review by sjoshisupra — CHANGES_REQUESTED** [review]
If schedule is 2/10, 3/10, 1/10, then add two test cases
1. Directly fast forward to a point from the beginning where 5/10 would be vested
2. Directly fast forward to a case from the beginning where 7/10 would be vested
In both cases above that exactly needed amount is vested and nothing more.

File: aptos-move/framework/aptos-stdlib/sources/fixed_point64.move (hunk 3 above)
Comment (ghost) [reply aptos-move/framework/aptos-stdlib/sources/fixed_point64.move]: I think we don't even need it right now as we did not use this anymore. Do we need to keep this function?

**Comment by ghost** [comment]
> If schedule is 2/10, 3/10, 1/10, then add two test cases
> 
> 1. Directly fast forward to a point from the beginning where 5/10 would be vested
> 2. Directly fast forward to a case from the beginning where 7/10 would be vested
>    In both cases above that exactly needed amount is vested and nothing more.

Test cases are added. 5/10 needs an addition of 2 and 7/10 needs an addition of 3 for the assertions. This behavior is also consistent with dev branch.

File: aptos-move/framework/supra-framework/sources/vesting_without_staking.move [hunk 4]
```diff
@@ -1604,4 +1631,209 @@ module supra_framework::vesting_without_staking {
...
+        // Because the time is behind the start time, vest will do nothing.
+        vest(contract_address);
+        assert!(coin::balance<SupraCoin>(contract_address) == 1000, 0);
+        assert!(coin::balance<SupraCoin>(shareholder_address) == vested_amount, 0);
+        // Time is now at the start time, vest will unlock the first period, which is 2/10.
+        timestamp::update_global_time_for_test_secs(vesting_start_secs(contract_address) + period_duration_secs(
+            contract_address) * 2);
+        vest(contract_address);
+        vested_amount = vested_amount + fraction(shareholder_share, 5, 10);
+        assert!(coin::balance<SupraCoin>(shareholder_address) + 2 == vested_amount, 0);
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/vesting_without_staking.move:1801]: Why is this `+ 2 ` required in assert? @axiongsupra

File: aptos-move/framework/supra-framework/sources/vesting_without_staking.move [hunk 5]
```diff
@@ -1604,4 +1631,209 @@ module supra_framework::vesting_without_staking {
...
+        // Because the time is behind the start time, vest will do nothing.
+        vest(contract_address);
+        assert!(coin::balance<SupraCoin>(contract_address) == 1000, 0);
+        assert!(coin::balance<SupraCoin>(shareholder_address) == vested_amount, 0);
+        // Time is now at the start time, vest will unlock the first period, which is 2/10.
+        timestamp::update_global_time_for_test_secs(vesting_start_secs(contract_address) + period_duration_secs(
+            contract_address) * 4);
+        vest(contract_address);
+        vested_amount = vested_amount + fraction(shareholder_share, 7, 10);
+        assert!(coin::balance<SupraCoin>(shareholder_address) + 3 == vested_amount, 0);
```
Comment (sjoshisupra) [inline aptos-move/framework/supra-framework/sources/vesting_without_staking.move:1837]: @axiongsupra why is this `+ 3` required here? Given that `2/10, 3/10, 1/10` can multiply perfectly with `1000`.

File: aptos-move/framework/supra-framework/sources/vesting_without_staking.move (hunk 4 above)
Comment (ghost) [reply aptos-move/framework/supra-framework/sources/vesting_without_staking.move:1801]: @sjoshisupra The actual vested value for each iteration is 199, 299. So in total would be 498 while the calculated vested amount is 500.

File: aptos-move/framework/supra-framework/sources/vesting_without_staking.move (hunk 5 above)
Comment (ghost) [reply aptos-move/framework/supra-framework/sources/vesting_without_staking.move:1837]: The actual vested value is 199, 299, 99, 99 so in total is 696 while the calculated vested amount is 699 in ` vested_amount = vested_amount + fraction(shareholder_share, 7, 10)`

**Review by sjoshisupra — APPROVED** [review]

**Pull request merged**